sdk/python/src/dagger/telemetry/__init__.py

```diff
@@ -0,0 +1,185 @@
+    """
+    name = (
+        f"OTEL_PYTHON_{name.upper()}_INSTRUMENTATION_DISABLED"
+        if name
+        else OTEL_SDK_DISABLED
+    )
+    return os.getenv(name, "").strip().lower() != "true"
+
+
+class _BaseConfigurator(_BaseSDKConfigurator):
+    """Base configurator singleton, that ensures configuration only happens once."""
+
+    _is_configured: bool = False
+
+    def configure(self, **kwargs):
+        if self._is_configured:
+            return
+
+        super().configure(**kwargs)
+        self._is_configured = True
+
+
+class _DaggerPropagationConfigurator(_BaseConfigurator):
+    # NB: This configuration should be applied before any other telemetry
+    # code runs, to ensure the context has the right traceparent.
+    def _configure(self, **kwargs):
+        if parent := os.getenv("TRACEPARENT"):
+            if propagation.get_current_span().get_span_context().is_valid:
+                return
+
+            logger.debug("Found TRACEPARENT", extra={"value": parent})
+            ctx = propagate.extract({"traceparent": parent})
+            context.attach(ctx)
+
+
+def _init_tracing(exporters: dict[str, type[SpanExporter]]):
+    # By default this is a NoOpTracerProvider, unless OTEL_PYTHON_TRACER_PROVIDER
+    # is set, which is done in _prepare_env.
+    provider = get_tracer_provider()
+
+    if isinstance(provider, TracerProvider):
+        for exporter_class in exporters.values():
+            # TODO: Use a LiveSpanProcessor (TBD).
+            provider.add_span_processor(BatchSpanProcessor(exporter_class()))
+
+
+class _DaggerOtelConfigurator(_BaseConfigurator):
+    # NB: This is based on opentelemetry.sdk._configuration._OtelSDKConfigurator
+    # which is experimental. Instead of importing just the configurator, we're
+    # importing several private functions because we need more control over
+    # the initialization of tracing exporters but still want to reuse as
+    # much of the existing logic as possible. Need to keep an eye on upstream
+    # changes though.
+    def _configure(self, **kwargs):
+        if not otel_configured():
+            logger.debug("Telemetry not configured")
+            return
+
+        if not otel_enabled():
+            logger.debug("Telemetry disabled")
+            return
+
+        logger.debug("Initializing telemetry")
+        self._prepare_env()
+        self._initialize()
+        logger.debug("Telemetry initialized")
+
+    def _prepare_env(self):
+        """Prepare environment variables for auto-configuring the SDK."""
+        # When a Resource is created, it defaults to the following env var
+        # for the service name.
+        os.environ.setdefault(OTEL_SERVICE_NAME, SERVICE_NAME)
+
+        # The default is a NoOpProvider.
+        os.environ.setdefault(OTEL_PYTHON_TRACER_PROVIDER, "sdk_tracer_provider")
+
+        # TODO: The INSECURE env vars should be set by the shim next to ENDPOINT
```

Comment:
Agreed, ran into this with custom Go integration for tests, don't remember if I made this change yet (on mobile)